### PR TITLE
Pass BrowserContext tests

### LINKED ISSUE #51: Pass browsercontext.spec.js tests in Chromium
Let's turn these tests green, let's remember to check Puppeteer's code

### PR DESCRIPTION
**Pull request by kblok**
relates to #51

### CODE REVIEW

File: src/PlaywrightSharp.Chromium/ChromiumBrowser.cs [hunk 1]
```diff
@@ -39,35 +37,46 @@ internal ChromiumBrowser(IBrowserApp app, ChromiumConnection connection, string[
...
 
-        /// <inheritdoc cref="IBrowser"/>
-        public IBrowserContext[] BrowserContexts => null;
+        /// <inheritdoc cref="IBrowser.BrowserContexts"/>
+        public IEnumerable<IBrowserContext> BrowserContexts
+        {
+            get
+            {
+                yield return DefaultContext;
+
```
Comment (Meir017) [inline src/PlaywrightSharp.Chromium/ChromiumBrowser.cs]: remove extra line

File: src/PlaywrightSharp.Chromium/ChromiumBrowser.cs [hunk 2]
```diff
@@ -168,12 +177,12 @@ public Task<string> StopTracingAsync()
 
         internal static async Task<IBrowser> ConnectAsync(IBrowserApp app, ConnectOptions options)
         {
-            var transport = await BrowserHelper.CreateTransportAsync(options).ConfigureAwait(false);
-            var connection = new ChromiumConnection(transport);
-            var response = await connection.RootSession.SendAsync(new TargetGetBrowserContextsRequest()).ConfigureAwait(false);
-            var browser = new ChromiumBrowser(app, connection, response.BrowserContextIds);
-            await connection.RootSession.SendAsync(new TargetSetDiscoverTargetsRequest { Discover = true }).ConfigureAwait(false);
-            await browser.WaitForTargetAsync(t => t.Type == TargetType.Page).ConfigureAwait(false);
+            var transport = await BrowserHelper.CreateTransportAsync(options: options).ConfigureAwait(continueOnCapturedContext: false);
```
Comment (Meir017) [inline src/PlaywrightSharp.Chromium/ChromiumBrowser.cs]: do we really need the parameters names everywhere here?

File: src/PlaywrightSharp.Chromium/ChromiumBrowserContext.cs [hunk 3]
```diff
@@ -42,7 +42,7 @@ internal ChromiumBrowserContext(ChromiumSession client, ChromiumBrowser chromium
         /// <inheritdoc cref="IBrowserContextDelegate.BrowserContext"/>
         public BrowserContext BrowserContext { get; set; }
 
-        internal ChromiumBrowser Browser { get; }
+        private ChromiumBrowser Browser { get; }
```
Comment (Meir017) [inline src/PlaywrightSharp.Chromium/ChromiumBrowserContext.cs]: could this be a field?

File: src/PlaywrightSharp.Chromium/ChromiumBrowser.cs (hunk 2 above)
Comment (kblok) [reply src/PlaywrightSharp.Chromium/ChromiumBrowser.cs]: ouch, I'm trying Rider 😂

File: src/PlaywrightSharp.Chromium/ChromiumExecutionContext.cs [hunk 4]
```diff
@@ -76,7 +76,19 @@ public async Task<T> EvaluateAsync<T>(FrameExecutionContext context, bool return
 
         public Task ReleaseHandleAsync(JSHandle handle) => ReleaseObjectAsync(_client, handle.RemoteObject);
 
-        private static object ValueFromUnserializableValue(RemoteObject remoteObject, string unserializableValue)
+        public string HandleToString(IJSHandle handle, bool includeType)
```
Comment (Meir017) [inline src/PlaywrightSharp.Chromium/ChromiumExecutionContext.cs]: could we also override the native `ToString` method?

File: src/PlaywrightSharp.Chromium/ChromiumPage.cs [hunk 5]
```diff
@@ -251,6 +287,94 @@ private void Client_MessageReceived(object sender, IChromiumEvent e)
...
+
+            try
+            {
+                await Task.WhenAll(
+                    session.SendAsync(new RuntimeEnableRequest()),
+                    session.SendAsync(new NetworkEnableRequest())).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                System.Diagnostics.Debug.WriteLine(ex);
```
Comment (Meir017) [inline src/PlaywrightSharp.Chromium/ChromiumPage.cs]: shouldn't we throw or something?

Comment (kblok) [reply src/PlaywrightSharp.Chromium/ChromiumPage.cs]: We should translate `.catch(debugError);` to error logging, but not throwing.

Comment (Meir017) [reply src/PlaywrightSharp.Chromium/ChromiumPage.cs]: and in the future `logger.LogDebug` ;)

File: src/PlaywrightSharp.Chromium/ChromiumBrowser.cs (hunk 1 above)
Comment (kblok) [reply src/PlaywrightSharp.Chromium/ChromiumBrowser.cs]: I don't consider this an extra line. I like the separation between the yield and the foreach.

File: src/PlaywrightSharp.Chromium/ChromiumExecutionContext.cs (hunk 4 above)
Comment (kblok) [reply src/PlaywrightSharp.Chromium/ChromiumExecutionContext.cs]: We can't do that. It's a browser implementation.

**Comment by kblok** [comment]
Ready for a final round @Meir017

File: src/PlaywrightSharp/Frame.cs [hunk 6]
```diff
@@ -174,9 +185,45 @@ public Task<T> QuerySelectorEvaluateAsync<T>(string selector, string script, par
...
+                // Clear lifecycle right after document.open() - see 'tag' below.
+                Page.FrameManager.ClearFrameLifecycle(this);
+                watcher = new LifecycleWatcher(this, options);
+            });
+
+            await context.EvaluateAsync(
+                @"(html, tag) => {
+                    window.stop();
+                    document.open();
+                    console.debug(tag);  // eslint-disable-line no-console
```
Comment (Meir017) [inline src/PlaywrightSharp/Frame.cs]: we don't need the eslint comment...

File: src/PlaywrightSharp/Helpers/PlaywrightSharpExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+using System.Linq;
+
+namespace PlaywrightSharp.Helpers
+{
+    internal static class PlaywrightSharpExtensions
+    {
+        internal static string GetText(this ConsoleMessage message)
```
Comment (Meir017) [inline src/PlaywrightSharp/Helpers/PlaywrightSharpExtensions.cs]: consider changing `ConsoleMessage.Text` to internal

**Pull request merged**